Add remaining UML elements

vpchung opened the pull request:
In addition to adding operations for creating and patching a challenge:

- added ROCC logos to docs page
- removed Health endpoints and operations (not needed for ROCC)
- added examples to some of the operations
- make `id` parameters `readOnly` and `password` parameter `writeOnly`

tschaffter commented on openapi/components/schemas/Challenge.yaml:
```diff
@@ -2,42 +2,61 @@ type: object
+    type: array
+    items:
+      $ref: Grant.yaml
+  organizers:
```
I would rename this property `organizations`

vpchung replied:
This is actually something I wanted to clarify.

For a challenge, do we want to name the specific people from organizations that are involved in the challenge i.e. what we are currently doing on Synapse? Or should we list the entire organization(s) as our "Partners" (similar to what Kaggle is doing)?

tschaffter replied:
What about:
- Have Challenge.organizers that is a list of Persons
- Have Challenge.organizations that is a list of Organizations. I should keep the name of the property based on the name of the schema used (here Organization).

This comment also suggest to distinguish a Person and User. Depending on the approach we take, Kaggle could be a User.

vpchung replied:
I decided to do:
- `Challenge.organizers`, where it is a list of Users
- add a relationship between User and Organization by adding `User:organization`

This more closely resembles what the current landscape is providing (and also reflects the UML that I designed). We can definitely restructure/refactor at a later time if needed!

tschaffter replied:
> Challenge.organizers, where it is a list of Users

This means that we are enforcing that a person must register as a User of ROCC in order to be able to be listed by someone else as an organizer of a given challenge. This is approach would prevent a platform like Kaggle to automatically register/synchronize challenges on ROCC.

Alternatively, we could have a Person object that has a `name` and `email`. Persons objects can easily be created by a challenge platform. We could use Person objects in the list of challenge organizers.

Imagine the challenge page. We could display the Person information as plain text by default. Now if someone has created a User account on the ROCC portal, we could use the `email` address as an identifier to link the Person and User object. Back the the challenge page, we could then imagine to make the name of challenge organizers who have a ROCC account clickable to go to the profile page of the User selected.

vpchung replied:
That makes sense! So essentially move forward with this design?

<img width="302" alt="Screen Shot 2020-11-12 at 8 51 06 PM" src="https://user-images.githubusercontent.com/9377970/99030001-d22cc100-2528-11eb-93bf-12503daefdb6.png">

I thought from our last discussion, we didn't want to move forward with this idea, but perhaps I misunderstood.

tschaffter replied:
Yes, considering the comments below:

- It's a good coding practice to fully spell variable names to improve code readability (`fname` => `firstName`)
- User
  - remove `organizer` and `admin` as discussed previously
  - add `role`
  - add `organization` (the principal org the User belongs to)

vpchung replied:
I added a `Person` model, keeping the criteria mentioned above.

tschaffter commented on openapi/components/schemas/Grant.yaml:
```diff
@@ -1,6 +1,10 @@
 description: Information about monetary resources for challenge
 properties:
+  id:
+    type: integer
```
Change type to string (MongoDB)

tschaffter commented on openapi/components/schemas/Challenge.yaml:
```diff
@@ -2,42 +2,61 @@ type: object
+  status: closed
+  tags: ["Machine Learning"]
+  grant: []
+  organizers: ["Organization1"]
```
- Rename property `organizations`
- Provide an example of Organization object, not a string (see https://swagger.io/docs/specification/adding-examples/)

tschaffter commented on openapi/components/schemas/Organization.yaml:
```diff
@@ -1,9 +1,13 @@
 properties:
+  id:
+    type: integer
+    format: int32
```
With string type, `uuid` could be an option but it's not an official OpenAPI value. Mongodb ObjectId are also not technically UUID by default (https://en.wikipedia.org/wiki/Universally_unique_identifier). I would just remove `format` for now and add it later if needed.

vpchung replied:
Ah, I played around with `id` a lot; some are `string: uuid`, some are `integer: int32`, and some `string: pattern` :sweat_smile:

I'll stick with a simple `string` for now

tschaffter commented on openapi/components/schemas/User.yaml:
```diff
@@ -1,10 +1,18 @@
+  id:
+    type: integer
+    format: int32
+    readOnly: true
```
This is great! This will solve a problem I was facing in another API.

tschaffter commented on openapi/components/schemas/User.yaml:
```diff
@@ -1,10 +1,18 @@
     type: string
     minLength: 4
     example: John78
+  password:
```
Add description that says that this field contain the password hash.

tschaffter replied:
Add a string property named `salt` (https://en.wikipedia.org/wiki/Salt_(cryptography))

vpchung replied:
Adding this to TODO -- add salt hashing for password

tschaffter commented on openapi/components/schemas/User.yaml:
```diff
@@ -17,8 +25,14 @@ properties:
     example: Smith
   email:
     $ref: Email.yaml
+  organizer:
```
How would this property be used?

vpchung replied:
This is a flag I'm using to determine whether an operation is allowed by a user, e.g. only organizers and admins can add/update/remove a challenge.

I don't know if we'll ever need an `admin` user-level, so that property can be removed later.

tschaffter replied:
I would add a property `role`, which is an enum that can take the value `admin` and `user` for now.

- `admin`: can do anything
- `user`: can create Challenges, Organizations, etc. and manage the entities they own

vpchung replied:
Oh, I like that!

vpchung commented:
Since the last review:
- changed all `id` properties to be a simple `string` (for now)
- updated the item-type of `organizers` to be User instead of Organization
- related above, added an `organization` property to the User model

tschaffter commented:
> updated the item-type of organizers to be User instead of Organization

See above comment

tschaffter reviewed (commented):
See comments added to the discsussion

tschaffter reviewed (approved):
👍

vpchung commented:
Having a hard time signing the previous commits.... so I'm closing this PR and opening another.
